pkg/controller/seed-controller-manager/mla/rulegroup_controller.go

```diff
@@ -0,0 +1,353 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mla
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	kubermaticv1helper "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1/helper"
+	"k8c.io/kubermatic/v2/pkg/kubernetes"
+	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ruleGroupFinalizer             = "kubermatic.io/rule-group"
+	metricsRuleGroupConfigEndpoint = "/api/v1/rules"
+	ruleGroupTenantHeaderName      = "X-Scope-OrgID"
+	defaultNamespace               = "/default"
+)
+
+type ruleGroupReconciler struct {
+	ctrlruntimeclient.Client
+	log                 *zap.SugaredLogger
+	workerName          string
+	recorder            record.EventRecorder
+	versions            kubermatic.Versions
+	ruleGroupController *ruleGroupController
+}
+
+func newRuleGroupReconciler(
+	mgr manager.Manager,
+	log *zap.SugaredLogger,
+	numWorkers int,
+	workerName string,
+	versions kubermatic.Versions,
+	ruleGroupController *ruleGroupController,
+) error {
+	log = log.Named(ControllerName)
+	client := mgr.GetClient()
+
+	reconciler := &ruleGroupReconciler{
+		Client:              client,
+		log:                 log,
+		workerName:          workerName,
+		recorder:            mgr.GetEventRecorderFor(ControllerName),
+		versions:            versions,
+		ruleGroupController: ruleGroupController,
+	}
+
+	ctrlOptions := controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	}
+	c, err := controller.New(ControllerName, mgr, ctrlOptions)
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.RuleGroup{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to watch RuleGroup: %w", err)
+	}
+
+	enqueueRuleGroupsForCluster := handler.EnqueueRequestsFromMapFunc(func(object ctrlruntimeclient.Object) []reconcile.Request {
+		cluster := object.(*kubermaticv1.Cluster)
+		if cluster.Status.NamespaceName == "" {
+			return nil
+		}
+		ruleGroupList := &kubermaticv1.RuleGroupList{}
+		if err := client.List(context.Background(), ruleGroupList, ctrlruntimeclient.InNamespace(cluster.Status.NamespaceName)); err != nil {
+			log.Errorw("failed to list ruleGroups for cluster", zap.Error(err), "cluster", cluster.Name)
```

Comment:
the error will be also logged by `HandleError`

Reply:
`HandleError` uses klog underneath, and I am not familiar it at all. Maybe let's just log it with our own zap logger, and we can make sure this is logged when bad thing happens here.